src/main/scala/temple/DSL/semantics/Analyser.scala

```diff
@@ -0,0 +1,264 @@
+        )
+      case "date" =>
+        assertNoParameters(args)
+        AttributeType.DateType
+      case "datetime" =>
+        assertNoParameters(args)
+        AttributeType.DateTimeType
+      case "time" =>
+        assertNoParameters(args)
+        AttributeType.TimeType
+      case "bool" =>
+        assertNoParameters(args)
+        AttributeType.BoolType
+      case typeName => fail(s"Unknown type $typeName")
+    }
+  }
+
+  def parseAttribute(
+    dataType: syntax.AttributeType,
+    annotations: Seq[syntax.Annotation]
+  )(implicit keyNameContext: KeyName): Attribute = {
+    var accessAnnotation: Option[Annotation.AccessAnnotation] = None
+
+    def setAccessAnnotation(annotation: Annotation.AccessAnnotation): Unit =
+      accessAnnotation.fold { accessAnnotation = Some(annotation) } { existingAnnotation =>
+        fail(
+          s"Two scope annotations found for ${keyNameContext.keyName}: " +
+          s"${annotation.render} is incompatible with ${existingAnnotation.render}"
+        )
+      }
+    val valueAnnotations = mutable.HashSet[Annotation.ValueAnnotation]()
+    annotations.iterator.map(_.key) foreach {
+      case "unique"    => valueAnnotations += Annotation.Unique
+      case "server"    => setAccessAnnotation(Annotation.Server)
+      case "client"    => setAccessAnnotation(Annotation.Client)
+      case "serverSet" => setAccessAnnotation(Annotation.ServerSet)
+      case key         => fail(s"Unknown annotation @$key at ${keyNameContext.keyName}")
+    }
+    Attribute(parseAttributeType(dataType), accessAnnotation, valueAnnotations.toSet)
+  }
+
+  /** A parser of Metadata items that can occur in all the blocks */
+  private val parseGlobalMetadata = new MetadataParser[GlobalMetadata] {
+    registerKeyword[String]("language", TokenArgType)(Language.parse(_))
+  }
+
+  /** A parser of Metadata items that can occur in project and service blocks */
+  private val parseProjectAndServiceMetadata = new MetadataParser[ProjectAndServiceMetadata] {
+    inherit from parseGlobalMetadata
+    registerKeyword("database", TokenArgType)(Database.parse(_))
+  }
+
+  /** A parser of Metadata items that can occur in service blocks */
+  private val parseServiceMetadata = new MetadataParser[ServiceMetadata] {
+    inherit from parseProjectAndServiceMetadata
+    registerKeyword("auth", "login", TokenArgType)(ServiceAuth)
+    registerKeyword("uses", "services", ListArgType(TokenArgType))(Uses)
+  }
+
+  /** A parser of Metadata items that can occur in project blocks */
+  private val parseProjectMetadata = new MetadataParser[ProjectMetadata] {
+    inherit from parseProjectAndServiceMetadata
+    registerKeyword("provider", TokenArgType)(Provider.parse(_))
+  }
+
+  /** A parser of Metadata items that can occur in target blocks */
+  private val parseTargetMetadata = new MetadataParser[TargetMetadata] {
+    inherit from parseGlobalMetadata
+    registerKeyword("auth", "services", ListArgType(TokenArgType))(TargetAuth)
```

Comment:
Personally, I don't like the `inherit from` here. Feels a bit like spaghetti code.
I think it'd make more sense to avoid the indirection and explicitly register each keyword with the target/project/service metadata parser every time